pyro/infer/predictive.py

```diff
@@ -219,3 +231,82 @@ def get_vectorized_trace(self, *args, **kwargs):
                                             parallel=self.parallel, model_args=args, model_kwargs=kwargs)
         return _predictive(self.model, posterior_samples, self.num_samples,
                            return_trace=True, model_args=args, model_kwargs=kwargs)
+
+
+def log_likelihood(
+    model: Callable,
+    posterior_samples: Dict[str, torch.Tensor] = None,
+    guide: Callable = None,
+    num_samples: int = None,
+    parallel: bool = True,
+) -> Callable:
+    """EXPERIMENTAL utility to calculate the log likelihood of the data for a
+    model with observed sites, given posterior samples (or a guide that can
+    generate them).
+
+    :param model: Model function containing pyro primitives. Must have observed
+        sites conditioned on data.
+    :type model: Callable
+    :param posterior_samples: Dictionary of posterior samples, defaults to None
+    :type posterior_samples: Dict[str, torch.Tensor], optional
+    :param guide: Guide function, defaults to None
+    :type guide: Callable, optional
+    :param num_samples: Number of posterior samples to evaluate, defaults to None
+    :type num_samples: int, optional
+    :param parallel: Whether to use vectorization, defaults to True
+    :type parallel: bool, optional
+    :rtype: Callable
+
+    The API parallels that of the :class:`Predictive` interface (assume `model`
+    and `guide` have already been defined elsewhere):
+
+    ```
+    x = torch.randn(10)
+    y = x + 0.05 * torch.randn(10)
+    cond_model = pyro.condition(model, data={"y": y})
+    ll_vals = log_likelihood(cond_model, guide=guide, num_samples)(x)
+    ```
+    """
+
+    _validate_predictive_inputs(posterior_samples, num_samples, guide, return_sites=None)
+
+    # define the function to return the likelihood values
+    def log_like_helper(*args, **kwargs) -> torch.Tensor:
+        # trace model once to get observed nodes
+        trace = pyro.poutine.trace(model).get_trace(*args, **kwargs)
+        observed = {
+            name: site["fn"].batch_shape
+            for name, site in trace.nodes.items()
+            if site["type"] == "sample" and site["is_observed"]
+        }
+        if len(observed) == 0:
+            raise RuntimeError("No observed sites in the model")
+        # use vectorized trace if parallelizable
+        if parallel:
+            log_like = dict()
+            # now trace it and extract the likelihood from observed sites
+            predictive = Predictive(
+                model, posterior_samples, guide, num_samples, (), parallel)
+            trace = predictive.get_vectorized_trace(*args, **kwargs)
+            for obs_name in observed.keys():
+                site = trace.nodes[obs_name]
+                log_like[obs_name] = site["fn"].log_prob(site["value"]).detach().cpu()
```

Comment:
nit: Probably we don't need to detach here (in NumPyro, we use `log_likelihood` in some algorithms that require taking `gradient`). But you can keep it and and a note for it in the docs and mention that this behavior might be changed in the future.